prow/plugins/branch-updater/branch-updater.go

```diff
@@ -0,0 +1,148 @@
+package branch_updater
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	gitHubContextNameTide        = "tide"
+	gitHubMergeableStateBehind   = "behind"
+	gitHubMergeStateRefreshDelay = 10
+	pluginName                   = "branch-updater"
+	tideContextStatusSuccess     = "success"
+)
+
+var (
+	enabledEvents = []github.PullRequestEventAction{
+		github.PullRequestActionOpened,
+		github.PullRequestActionEdited,
+		github.PullRequestActionReadyForReview,
+		github.PullRequestActionReopened,
+	}
+)
+
+type githubClient interface {
+	UpdatePullRequestBranch(owner, repo string, number int, expectedHeadSha *string) error
+	GetPullRequest(owner, repo string, number int) (*github.PullRequest, error)
+	GetCombinedStatus(org, repo, ref string) (*github.CombinedStatus, error)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []config.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	configInfo := map[string]string{}
+	for _, repo := range enabledRepos {
+		repoUpdateTmpl := "PR branches in repo %s will not be automatically updated."
+		if config.BranchUpdater.ShouldUpdateBranchesForRepo(repo.String()) {
+			repoUpdateTmpl = "PR branches in repo %s will be automatically updated."
+		}
+		configInfo[repo.String()] = fmt.Sprintf(repoUpdateTmpl, repo)
+	}
+
+	yamlSnippet, _ := plugins.CommentMap.GenYaml(&plugins.Configuration{
+		BranchUpdater: plugins.BranchUpdater{
+			DefaultEnabled: true,
+			IgnoredRepos:   []string{"db-improbable/do-not-update-me"},
+			IncludeRepos:   []string{"db-improbable/please-update-me"},
+		},
+	})
+
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The branch-updater plugin updates the branch when an otherwise-mergeable PR is blocked from merging because it's out of date from the HEAD branch",
+		Config:      configInfo,
+		Snippet:     yamlSnippet,
+	}
+	return pluginHelp, nil
+}
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, pullRequestHandler, helpProvider)
+}
+
+func pullRequestHandler(pc plugins.Agent, event github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.Logger, pc.PluginConfig.BranchUpdater, &event)
+}
+
+// If Tide thinks this PR should be mergeable, but GitHub says that the PR's branch is "behind",
+// ask GitHub to update the branch for us.
+func handlePR(gc githubClient, log *logrus.Entry, config plugins.BranchUpdater, event *github.PullRequestEvent) error {
+	// We only care about certain events, so ignore others - this significantly limits the number of race conditions
+	// that can cause multiple actions
+	relevantEvent := false
+	for _, candidate := range enabledEvents {
+		if event.Action == candidate {
+			relevantEvent = true
+		}
+	}
+
+	if !relevantEvent {
+		return nil
+	}
```

Comment:
worth extracting into a separate function to shortcut and make this easier to read?